text/0000-ssr-stylesheet-deduplication.md

```diff
@@ -0,0 +1,165 @@
+
+## Detailed design
+
+### Options
+
+At a high level, there are multiple options on the table:
+
+1. Bet on the native browser implementation and use it or a polyfill.
+2. Detect unused CSS rules on the server side (i.e. based on the rendered CSS selectors and HTML) and remove them.
+3. Extract duplicate stylesheets into `<link rel=stylesheet>`s and host the individual `.css` files externally.
+4. Use an inline `<script>` or web component to read styles from a global location (e.g. the `<head>`) and copy into individual shadow roots.
+
+After researching, especially after observing benchmark results, we've decided on option #4. The main reasons are:
+
+1. It is the fastest option by far. Depending on the browser, it is anywhere from 20% to 100% faster than `<link rel=stylesheet>`s, the runner-up choice.
+2. It would work for both shadow DOM and light DOM components.
+3. It does not require waiting for browsers to implement a new API.
+4. It can operate transparently – by removing both the `<style>` and itself, the `<lwc-style>` component can avoid virtual DOM-diffing or hydration mismatch issues by more closely matching what a pure Client-Side-Rendered LWC component would do (i.e. use `adoptedStyleSheets` or hoist `<style>`s into the `<head>`.
+5. It avoids any issues that removing "unused" CSS rules may incur (e.g. if a rule ends up being used dynamically later on the client side).
+6. Compared to a `<script>`, it results in less duplicated content, and also avoids issues with cross-origin, CSP nonces, etc. The consumer is expected to define the web component themselves once, in the `<head>` preferably.
+
+### Design
+
+At a high level, we need to give users of the `renderComponent` API a way to specify which stylesheets already exist on the page, and which can be considered by `<lwc-style>` to be available for reuse.
+
+There were various options here, including adding a hook for generating HTML or virtual DOM, or even just asking consumers to use a regex on the generated HTML.
+
+The main complexity here is the need to render something different on the first vs subsquent encounter of a given stylesheet, and to allow reuse across multiple islands. The `Set` neatly solves both cases, as it communicates very clearly whether a stylesheet was already used or not.
+
+### API shape
+
+A new property bag is added to the `renderComponent` API of `@lwc/ssr-runtime`.
+
+```ts
+function renderComponent(
+    tagName: string, 
+    Constructor: LightningElement, 
+    props: any,
+    // The below is new
+    options?: {
+        stylesheetCache?: Set<String>;
+    }
+): void;
+```
+
+For the client side, consumers are expected to import the `registerLwcStyleComponent` function and call it once:
+
+```js
+import { registerLwcStyleComponent } from '@lwc/ssr-runtime'
+registerLwcStyleComponent()
+```

```

Comment:
What happens if `registerLwcStyleComponent()` is not called? Will all SSRed components be un-styled?

Reply:
Correct.

Reply:
Should this call occur before `hydrateComponent`?

Reply:
For sure, yes.